feat(sql): add approx_count_distinct() SQL function

piotrrzysko opened the pull request:
Resolves #3295.

I added support for int, long, and IPv4. I can extend it to other types that are already supported by `count_distinct()`, but first, I wanted to collect early feedback from you.

Additionally, the sparse implementation can be improved to consume less memory by introducing variable-length encoding.

bluestreak01 commented:
hey @piotrrzysko thank you for the contribution! This looks like a lot of work! I really appreciate it!

Our codebase moved on recently by implementing `count_distinct` using parallel execution. I tested your PR on my 32-core desktop and the difference between parallel and serial execution is about 40x.

It would be good idea to see how we can get your functions to execution concurrently. Perhaps @puzpuzpuz can help here. The goal would be to beat the exact count, otherwise approx does not make sense if you can get exact faster.

Makes sense?

piotrrzysko commented:
Thanks for the feedback. Sounds good. I'll work on adapting the functions for parallel execution.

puzpuzpuz commented:
@piotrrzysko first of all, many thanks for this contribution! It looks very solid.

I'm currently working on parallel execution support for `count_distinct()` and some other aggregate functions in #4097. Once (and hopefully) this PR lands, I'll ping you here and we'll discuss what can be done to make `approx_count_distinct()` parallel execution friendly.

puzpuzpuz commented:
> I added parallel execution to approx_count_distinct

This is awesome, thanks for going the extra mile with this change! I'm a bit busy today, so I'm going to review the PR at the beginning of the next week.

> I need to dig deeper to understand if my reasoning described in the TODO is correct.

`merge()` will be only called if the execution is parallel. So, it's not a concern as SAMPLE BY won't call it. Besides, when we make SAMPLE BY parallel, interpolation should only take place after merge, not before it - otherwise, the result won't be correct.

puzpuzpuz commented:
> I ran the following queries on data from ClickBench on my desktop (which has a very old CPU with only 4 cores)

Here is what I got on my 4c/8t machine:
```sql
-- execute: 1.7s
SELECT count_distinct(UserID) FROM hits;

-- execute: 342ms
SELECT approx_count_distinct(UserID) FROM hits;

-- execute: 2.63s
SELECT RegionID, count_distinct(UserID) AS u FROM hits GROUP BY RegionID ORDER BY u DESC LIMIT 10;

-- execute: 1.3s
SELECT RegionID, approx_count_distinct(UserID) AS u FROM hits GROUP BY RegionID ORDER BY u DESC LIMIT 10;

-- execute: 3.03s
SELECT RegionID, SUM(AdvEngineID), COUNT(*) AS c, AVG(ResolutionWidth), count_distinct(UserID) FROM hits GROUP BY RegionID ORDER BY c DESC LIMIT 10;

-- execute: 1.76s
SELECT RegionID, SUM(AdvEngineID), COUNT(*) AS c, AVG(ResolutionWidth), approx_count_distinct(UserID) FROM hits GROUP BY RegionID ORDER BY c DESC LIMIT 10;

-- execute: 272ms
SELECT MobilePhoneModel, count_distinct(UserID) AS u FROM hits WHERE MobilePhoneModel IS NOT NULL GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10;

-- execute: 207ms
SELECT MobilePhoneModel, approx_count_distinct(UserID) AS u FROM hits WHERE MobilePhoneModel IS NOT NULL GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10;

-- execute: 1.86s
SELECT SearchPhrase, count_distinct(UserID) AS u FROM hits WHERE SearchPhrase IS NOT NULL GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;

-- execute: 2.83s
SELECT SearchPhrase, approx_count_distinct(UserID) AS u FROM hits WHERE SearchPhrase IS NOT NULL GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;
```

Long story short, `approx_count_distinct()` is faster than `count_distinct()` everywhere except the last query. The full result set in that query is ~6M rows, so we use hash table sharding (a.k.a. radix partitioning). This means that the cores have to merge partial results (think, functions) a lot. Maybe, that's where `approx_count_distinct()` is slower?

@piotrrzysko WDYT?

puzpuzpuz commented:
@piotrrzysko sorry for the delay, there are a few other PRs I need to care of, so I need to switch periodically. I'll do my best to review your PR this week.

piotrrzysko commented:
No problem :). Thanks for running the queries. In the meantime, I'll try to address the comments you've already posted.

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/engine/functions/groupby/ApproxCountDistinctIPv4GroupByFunctionFactoryTest.java:
```diff
@@ -0,0 +1,232 @@
+import io.questdb.test.AbstractCairoTest;
+import org.junit.Test;
+
+public class ApproxCountDistinctIPv4GroupByFunctionFactoryTest extends AbstractCairoTest {
```
None of the tests are multi-threaded, i.e. the only test which tests merge is `HyperLogLogTest`. Would be great to add a few tests to `ParallelGroupByFuzzTest`, with both `approx_count_distinct()` and `count_distinct()` as aggregate functions (that's to make the test results easier to verify by a human).

piotrrzysko replied:
Done: 3ad1e59ac33ce7248b84281faf5c037d6e80cb79

puzpuzpuz replied:
All of the added tests seem to use sparse representation. Is that correct? If so, would be great to add a test that would hit dense representation.

piotrrzysko replied:
Done: 072dfb28ef4a88c147e83fc3c33b36f537126a7d

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/engine/groupby/hyperloglog/HyperLogLogTest.java:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    @Test
+    public void testHighCardinality() throws Exception {
```
Would be great to add a fuzz test similar to this one, but with randomized inputs and non-constant RNG seeds. See `Unordered8MapTest#testFuzz()` as an example.

piotrrzysko replied:
I like this approach! 

Addressed in 44dffdfcdd14ee8840f23ed23fea7705a3c8cf36. However, there is a small chance that the test I added  will fail. See the comment in `HyperLogLogTestUtils#assertCardinality`. I'm not sure if there is a way around it. I was considering testing the HLL++ from this PR against some other reference implementation. I can research this later.

puzpuzpuz replied:
We could relax the assertion with some kind of padding, e.g. 4 standard errors instead of 3. I don't think it's a big deal.

piotrrzysko commented on core/src/main/java/io/questdb/griffin/engine/functions/groupby/ApproxCountDistinctIPv4GroupByFunction.java:
```diff
@@ -0,0 +1,210 @@
+    public void pushValueTypes(ArrayColumnTypes columnTypes) {
+        setValueIndex(columnTypes.getColumnCount());
+        columnTypes.add(ColumnType.LONG);
+        columnTypes.add(ColumnType.BOOLEAN);
```
Done: ca7e8ff5dedbde6b4e9bd423425e65a32694707d

piotrrzysko commented on core/src/main/java/io/questdb/griffin/engine/functions/groupby/ApproxCountDistinctIPv4GroupByFunction.java:
```diff
@@ -0,0 +1,210 @@
+            if (srcCount == 0 || srcCount == Numbers.LONG_NaN) {
+                return;
+            }
+            // TODO: How to handle the case when srcCount is not equal to 0 and not equal to NaN? This can occur as a
```
Done: bb7abbe65a7825f1e4fac91cb012fce63d0ee4ba

piotrrzysko commented:
I addressed most of your comments. Now, I'll look into the performance issue with the last query. If I manage to fix it, I'll adress the rest of the comments.

puzpuzpuz commented:
> I addressed most of your comments. Now, I'll look into the performance issue with the last query. If I manage to fix it, I'll adress the rest of the comments.

It could be the case that murmur3 is slower than the function from the `Hash` class (I haven't profiled the code, just guessing). Also, I'm tweaking our hash function in #4161, so it'll get a bit faster than what we have now.

piotrrzysko commented:
I think I know why the performance of approx_distinct_count is worse for the last query. Here are flamegraphs captured while running the query in a loop (red rectangles mark `ApproxCountDistinctLongGroupByFunction.getLong` calls):

Cpu-cycles:
![cpu-cycles](https://github.com/questdb/questdb/assets/6481553/6fb01690-507e-493c-bfb0-f856545d43ba)

Bus-cycles:
![bus-cycles](https://github.com/questdb/questdb/assets/6481553/4f4cb657-0fa4-4b43-ae4e-1b1671508acb)

Cache-misses:
![cache-misses](https://github.com/questdb/questdb/assets/6481553/21c9fd42-8313-4ca4-8680-928f67014ff4)


For this particular query, cardinalities of each group are relatively small - mainly, the sparse implementation of HLL is used. Additionally, at the end, the estimated cardinalities are sorted. Even though the estimation is performed only once for each group (HLL has a cache), every time the values are compared, we have to jump to a memory location where the HLL instance for a given group is stored, which, of course, causes a lot of cache misses.

I played around with the code a bit and noticed that just calling the `HyperLogLog.of` function in `ApproxCountDistinctLongGroupByFunction.getLong` and skipping `HyperLogLog.computeCardinality` doesn’t change the performance much.

Perhaps a solution would be to store the size of sparse HLL in MapValue instead of in an “external” memory segment. I don’t expect, though, that beating `count_distinct` is possible in this case.

By the way, I wonder how useful this feature would be for the database customers. Even though, for most queries, `approx_distinct_count` is faster than `distinct_count`, the differences are not large. I’d say HLL has an advantage over HashSet only for specific cases: when the cardinality of an input set is high and the estimation is done infrequently. If you think it make sens to work on it, I'm happy to do so :).

piotrrzysko commented on core/src/main/java/io/questdb/griffin/engine/functions/groupby/ApproxCountDistinctIPv4GroupByFunction.java:
```diff
@@ -0,0 +1,210 @@
+
+    public ApproxCountDistinctIPv4GroupByFunction(Function arg) {
+        this.arg = arg;
+        this.hllA = new HyperLogLog();
```
Done: c567855521dd4e4e0c6119aebafbf91a98350462

piotrrzysko commented:
I think I addressed all the comments. 

I fixed the issue with cache misses: ffaacaaf73d512f2c25f06ef72bcccecbad9380e and reran the queries on my desktop:
```sql
-- execute: 2.31s
SELECT count_distinct(UserID) FROM hits;

-- execute: 748.6ms
SELECT approx_count_distinct(UserID) FROM hits;

-- execute: 3.05s
SELECT RegionID, count_distinct(UserID) AS u FROM hits GROUP BY RegionID ORDER BY u DESC LIMIT 10;

-- execute: 2.2s
SELECT RegionID, approx_count_distinct(UserID) AS u FROM hits GROUP BY RegionID ORDER BY u DESC LIMIT 10;

-- execute: 3.82s
SELECT RegionID, SUM(AdvEngineID), COUNT(*) AS c, AVG(ResolutionWidth), count_distinct(UserID) FROM hits GROUP BY RegionID ORDER BY c DESC LIMIT 10;

-- execute: 2.6s
SELECT RegionID, SUM(AdvEngineID), COUNT(*) AS c, AVG(ResolutionWidth), approx_count_distinct(UserID) FROM hits GROUP BY RegionID ORDER BY c DESC LIMIT 10;

-- execute: 241.25ms
SELECT MobilePhoneModel, count_distinct(UserID) AS u FROM hits WHERE MobilePhoneModel IS NOT NULL GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10;

-- execute: 148.31ms
SELECT MobilePhoneModel, approx_count_distinct(UserID) AS u FROM hits WHERE MobilePhoneModel IS NOT NULL GROUP BY MobilePhoneModel ORDER BY u DESC LIMIT 10;

-- execute: 5.63s
SELECT SearchPhrase, count_distinct(UserID) AS u FROM hits WHERE SearchPhrase IS NOT NULL GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;

-- execute: 5.12s
SELECT SearchPhrase, approx_count_distinct(UserID) AS u FROM hits WHERE SearchPhrase IS NOT NULL GROUP BY SearchPhrase ORDER BY u DESC LIMIT 10;
```

puzpuzpuz commented on core/src/test/java/io/questdb/test/griffin/engine/groupby/hyperloglog/HyperLogLogTest.java:
```diff
@@ -0,0 +1,361 @@
+    }
+
+    @Test
+    public void testFuzz() throws Exception {
```
This test takes 13+ secs on my machine. On CI, it'll be even slower. We need to speed it up. An option to do that might be to use random precision from the allowed range instead of looping through all of them.

piotrrzysko replied:
Replacing the loop with randomly selecting precision reduced the execution time from 15 seconds to 1 second on my desktop.

puzpuzpuz replied:
Thanks for the speed up.

puzpuzpuz reviewed (commented):
Epic job! LGTM

Would be great to speed up `testFuzz` before merging this one.

puzpuzpuz commented:
Congrats on this epic contribution 🥳
